rust_src/src/dispnew.rs

```diff
@@ -143,13 +143,34 @@ pub fn internal_show_cursor_p(window: LispWindowOrSelected) -> bool {
     !win.cursor_off_p()
 }
 
+/// Return whether input is coming from the keyboard.
+// Corresponds to the INTERACTIVE macro in commands.h.
+pub fn is_interactive() -> bool {
+    unsafe { globals.Vexecuting_kbd_macro.is_nil() && !noninteractive }
+}
+
```

Comment:
If we gave `ding_internal` an argument we could collapse the Lisp ding and ding_internal code. The only difference between them is whether to call when interactive but currently in a non-interactive mode.